File: src/plugins/mermaid.zig
const std = @import("std");
const plugin = @import("../plugin.zig");
const Editor = @import("../Editor.zig");
const Self = @This();

/// Mermaid plugin — insert mermaid diagram templates.
/// Templates for flowcharts, sequences, gantt, class diagrams.
/// Commands: :mermaid, :mermaid.flow, :mermaid.seq, :mermaid.gantt
pub fn pluginInfo(self: *Self) plugin.PluginInfo {
    _ = self;
    return .{ .name = "mermaid", .version = "0.1.0", .author = "LazyMD contributors", .description = "Mermaid diagram support" };
}
pub fn pluginInit(self: *Self, editor: *Editor) void {
    _ = self;
    _ = editor;
}
pub fn pluginDeinit(self: *Self) void {
    _ = self;
}
pub fn onEvent(self: *Self, event: *plugin.PluginEvent) void {
    _ = self;
    _ = event;
}
pub fn getCommands(self: *Self) []const plugin.CommandDef {
    _ = self;
    return &[_]plugin.CommandDef{
        .{ .name = "mermaid", .description = "Insert mermaid flowchart", .handler = &insertFlow },
        .{ .name = "mermaid.flow", .description = "Insert flowchart template", .handler = &insertFlow },
        .{ .name = "mermaid.seq", .description = "Insert sequence diagram", .handler = &insertSeq },
        .{ .name = "mermaid.gantt", .description = "Insert gantt chart", .handler = &insertGantt },
    };
}

fn insertTemplate(editor: *Editor, template: []const u8) void {
    const offset = editor.buffer.posToOffset(editor.cursor_row, editor.cursor_col);
    editor.buffer.insertSlice(offset, template) catch {
        editor.status.set("Failed to insert diagram", true);
        return;
    };
    editor.status.set("Mermaid diagram inserted", false);
}

fn insertFlow(event: *plugin.PluginEvent) void {
    insertTemplate(event.editor,
        \\```mermaid
        \\graph TD
        \\    A[Start] --> B{Decision}
        \\    B -->|Yes| C[Action]
        \\    B -->|No| D[End]
        \\```
        \\
    );
}

fn insertSeq(event: *plugin.PluginEvent) void {
    insertTemplate(event.editor,
        \\```mermaid
        \\sequenceDiagram
        \\    Alice->>Bob: Hello
        \\    Bob-->>Alice: Hi there
        \\    Alice->>Bob: How are you?
        \\    Bob-->>Alice: Great!
        \\```
        \\
    );
}

fn insertGantt(event: *plugin.PluginEvent) void {
    insertTemplate(event.editor,
        \\```mermaid
        \\gantt
        \\    title Project Timeline
        \\    dateFormat YYYY-MM-DD
        \\    section Phase 1
        \\    Task 1 :a1, 2024-01-01, 30d
        \\    Task 2 :after a1, 20d
        \\    section Phase 2
        \\    Task 3 :2024-02-20, 15d
        \\```
        \\
    );
}

test "mermaid plugin" {
    var p = Self{};
    try std.testing.expectEqualStrings("mermaid", p.pluginInfo().name);
    const cmds = p.getCommands();
    try std.testing.expectEqual(@as(usize, 4), cmds.len);
}

File: internal/plugins/mermaid.go
package plugins

import (
	"github.com/EME130/lazymd/internal/editor"
)

type MermaidPlugin struct{}

func (p *MermaidPlugin) Info() PluginInfo {
	return PluginInfo{
		Name:        "mermaid",
		Version:     "0.1.0",
		Author:      "LazyMD contributors",
		Description: "Mermaid diagram support",
	}
}

func (p *MermaidPlugin) Init(ed editor.PluginEditor) {}

func (p *MermaidPlugin) OnEvent(event *PluginEvent) {}

func (p *MermaidPlugin) Commands() []CommandDef {
	return []CommandDef{
		{Name: "mermaid", Description: "Insert mermaid diagram", Handler: mermaidHandler},
		{Name: "mermaid.flow", Description: "Insert flowchart template", Handler: mermaidFlowHandler},
		{Name: "mermaid.seq", Description: "Insert sequence diagram", Handler: mermaidSeqHandler},
		{Name: "mermaid.gantt", Description: "Insert gantt chart", Handler: mermaidGanttHandler},
	}
}

func mermaidHandler(ed editor.PluginEditor, args string) {
	buf := ed.Buffer()
	row := ed.CursorRow()
	col := ed.CursorCol()
	offset := buf.PosToOffset(row, col)
	buf.InsertString(offset, "```mermaid\ngraph TD\n    A-->B\n```\n")
	ed.SetStatus("Mermaid diagram inserted", false)
}

func mermaidFlowHandler(ed editor.PluginEditor, args string) {
	buf := ed.Buffer()
	row := ed.CursorRow()
	col := ed.CursorCol()
	offset := buf.PosToOffset(row, col)
	template := "```mermaid\nflowchart TD\n    Start --> Process\n    Process --> End\n```\n"
	buf.InsertString(offset, template)
	ed.SetStatus("Flowchart template inserted", false)
}

func mermaidSeqHandler(ed editor.PluginEditor, args string) {
	buf := ed.Buffer()
	row := ed.CursorRow()
	col := ed.CursorCol()
	offset := buf.PosToOffset(row, col)
	template := "```mermaid\nsequenceDiagram\n    Alice->>Bob: Hello\n    Bob->>Alice: Hi\n```\n"
	buf.InsertString(offset, template)
	ed.SetStatus("Sequence diagram inserted", false)
}

func mermaidGanttHandler(ed editor.PluginEditor, args string) {
	buf := ed.Buffer()
	row := ed.CursorRow()
	col := ed.CursorCol()
	offset := buf.PosToOffset(row, col)
	template := "```mermaid\ngantt\n    title Project Schedule\n    section Phase 1\n    Task 1 :a1, 2024-01-01, 30d\n```\n"
	buf.InsertString(offset, template)
	ed.SetStatus("Gantt chart inserted", false)
}
